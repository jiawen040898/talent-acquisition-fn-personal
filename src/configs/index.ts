export * from './auth.config';
export * from './aws.config';
export * from './skill-extract-api.config';
export * from './unleash.config';
