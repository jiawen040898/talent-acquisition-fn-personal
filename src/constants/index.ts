export * from './action-history.enum';
export * from './auth-error.enum';
export * from './cache-store.enum';
export * from './candidate.event-type.enum';
export * from './channel.enum';
export * from './contact-type.enum';
export * from './ds-lookup-api-errors.enum';
export * from './email-event.enum';
export * from './email-type.enum';
export * from './error-msg.enum';
export * from './errors';
export * from './feature-toggle-name.enum';
export * from './fit-score-recipe.enum';
export * from './identity-event.enum';
export * from './job.enum';
export * from './job-application-score-type.enum';
export * from './job-application-skill.enum';
export * from './job-application-status.enum';
export * from './person-score.enum';
export * from './person-score-event.enum';
export * from './psychology-api-errors.enum';
export * from './questionnaire-framework.enum';
export * from './s3-errors.enum';
export * from './screening-answer-criteria.enum';
export * from './skill-extract-api-errors.enum';
export * from './talent-acquisition-event.enum';
export * from './talent-management-event.enum';
export * from './timeline.enum';
export * from './unleash-errors.enum';
export * from './user.enum';
