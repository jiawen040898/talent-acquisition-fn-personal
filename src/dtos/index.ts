export * from './audit-data.dto';
export * from './candidate-event.dto';
export * from './create-job.dto';
export * from './fit-score-recipe.dto';
export * from './geolocation.dto';
export * from './identity-update-user-event.dto';
export * from './job-application.dto';
export * from './job-application-action-history.dto';
export * from './job-application-attachment.dto';
export * from './job-application-career.dto';
export * from './job-application-contact.dto';
export * from './job-application-education.dto';
export * from './job-application-questionnaire.dto';
export * from './job-application-resume.dto';
export * from './job-application-score.dto';
export * from './job-application-screening-answer.dto';
export * from './person-score-event.dto';
export * from './place.dto';
export * from './talent-acquisition-event.dto';
export * from './talent-management-event.dto';
