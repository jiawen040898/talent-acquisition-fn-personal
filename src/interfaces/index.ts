export * from './api.interface';
export * from './api-response.interface';
export * from './aws.interface';
export * from './candidate.interface';
export * from './cognitive-score.interface';
export * from './email.interface';
export * from './event-model.interface';
export * from './job-application-score.interface';
export * from './pairwise-score.interface';
export * from './personality.interface';
export * from './personality-score.interface';
export * from './skill-extract-api.interface';
export * from './user.interface';
