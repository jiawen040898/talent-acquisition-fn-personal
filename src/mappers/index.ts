export * from './job-application-mapper';
export * from './job-application-score-mapper';
