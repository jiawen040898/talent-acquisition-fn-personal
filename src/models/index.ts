export * from './audit-data.entity';
export * from './entities';
export * from './job.entity';
export * from './job-application.entity';
export * from './job-application-action-history.entity';
export * from './job-application-attachment.entity';
export * from './job-application-career.entity';
export * from './job-application-contact.entity';
export * from './job-application-education.entity';
export * from './job-application-label.entity';
export * from './job-application-questionnaire.entity';
export * from './job-application-resume.entity';
export * from './job-application-score.entity';
export * from './job-application-screening-answer.entity';
export * from './job-distribution-score.entity';
export * from './job-questionnaire.entity';
export * from './job-screening-question.entity';
export * from './job-user-access.entity';
