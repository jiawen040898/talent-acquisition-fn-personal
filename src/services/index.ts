export * from './beta-distribution.service';
export * from './cognitive-score.service';
export * from './ds-lookup.service';
export * from './feature-toggle.service';
export * from './fit-score.service';
export * from './job-application.service';
export * from './job-application-score.service';
export * from './job-application-skill-extraction.service';
export * from './job-distribution-score.service';
export * from './job-user-access.service';
export * from './pairwise-score.service';
export * from './personality-score.service';
export * from './psychology.service';
export * from './publisher.service';
export * from './s3.service';
export * from './skill-extract-api.service';
