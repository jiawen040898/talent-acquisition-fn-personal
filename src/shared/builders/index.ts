export * from './job-application-query.builder';
export * from './job-application-score-query.builder';
export * from './job-query.builder';
