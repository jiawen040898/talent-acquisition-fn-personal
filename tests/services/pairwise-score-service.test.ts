import {
    JobApplicationScoreType,
    JobApplicationSkillSource,
} from '@pulsifi/constants';
import {
    JobApplicationSkill,
    PairwiseScoreOutcomePayload,
} from '@pulsifi/interfaces';
import { Job, JobApplication, JobApplicationCareer } from '@pulsifi/models';
import { DSLookupService, PairwiseScoreService } from '@pulsifi/services';
import {
    testJobApplicationBuilder,
    testJobApplicationCareerBuilder,
    testJobBuilder,
} from '@pulsifi/tests/builders';
import { testJobApplicationSkillBuilder } from '@pulsifi/tests/builders/test-job-application-score-builder';
import { mockRoleFitRecipeWithCompetencyInclusiveness } from '@pulsifi/tests/fixtures';
import { getTestDataSource } from '@pulsifi/tests/setup';
import { DataSource } from 'typeorm';

import { getDataSource, setDataSource } from '../../src/database';

jest.mock('../../src/services/ds-lookup.service');

const mockGetPairwiseSimilarity = jest.fn();
DSLookupService.getPairwiseSimilarity = mockGetPairwiseSimilarity;
mockGetPairwiseSimilarity.mockResolvedValue({
    score: 0.8,
    pairwise_score: {
        Javascript: {
            Python: 0.7,
            Typescript: 0.9,
        },
    },
});

describe('PairwiseScoreService', () => {
    let dataSource: DataSource;
    let jobApplicationId: string;
    beforeAll(async () => {
        setDataSource(await getTestDataSource());

        dataSource = await getDataSource();

        jobApplicationId = '00000000-0000-0000-0001-000000000001';
        const job = testJobBuilder.build();
        const jobApplication = testJobApplicationBuilder.build({
            id: jobApplicationId,
            job_id: job.id,
        });
        const jobApplicationCareer = testJobApplicationCareerBuilder.build({
            job_application_id: jobApplication.id,
        });

        await dataSource.getRepository(Job).save(job);
        await dataSource.getRepository(JobApplication).save(jobApplication);
        await dataSource
            .getRepository(JobApplicationCareer)
            .save(jobApplicationCareer);
    });

    describe('getHardSkillJobApplicationScore', () => {
        it('should return the correct hard skill job application score', async () => {
            // Arrange
            const recipe = mockRoleFitRecipeWithCompetencyInclusiveness;
            const skills = [
                testJobApplicationSkillBuilder.build({
                    name: 'Javascript',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
                testJobApplicationSkillBuilder.build({
                    name: 'Python',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
                testJobApplicationSkillBuilder.build({
                    name: 'Typescript',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
            ];

            // Act
            const result =
                await PairwiseScoreService.getHardSkillJobApplicationScore(
                    skills,
                    recipe,
                    recipe.id,
                    jobApplicationId,
                    5,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should handle the case where the job application has no skills', async () => {
            // Arrange
            const recipe = mockRoleFitRecipeWithCompetencyInclusiveness;
            const skills: JobApplicationSkill[] = [];

            // Act
            const result =
                await PairwiseScoreService.getHardSkillJobApplicationScore(
                    skills,
                    recipe,
                    recipe.id,
                    jobApplicationId,
                    5,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should handle the case where recipe have no skills', async () => {
            // Arrange
            const recipe = Object.assign(
                {},
                mockRoleFitRecipeWithCompetencyInclusiveness,
            );
            recipe.job_competency = [];

            const skills = [
                testJobApplicationSkillBuilder.build({
                    name: 'Javascript',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
                testJobApplicationSkillBuilder.build({
                    name: 'Python',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
                testJobApplicationSkillBuilder.build({
                    name: 'Typescript',
                    source: JobApplicationSkillSource.DAXTRA,
                }),
            ];

            // Act
            const result =
                await PairwiseScoreService.getHardSkillJobApplicationScore(
                    skills,
                    recipe,
                    recipe.id,
                    jobApplicationId,
                    5,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('getWorkExpJobApplicationScore', () => {
        it('should return the correct work experience job application score', async () => {
            // Act
            const result =
                await PairwiseScoreService.getWorkExpJobApplicationScore(
                    jobApplicationId,
                    mockRoleFitRecipeWithCompetencyInclusiveness,
                    mockRoleFitRecipeWithCompetencyInclusiveness.id,
                    5,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should trim job if the job key provided have space', async () => {
            // Arrange
            mockGetPairwiseSimilarity.mockResolvedValueOnce({
                score: 0.8,
                pairwise_score: {
                    // eslint-disable-next-line @typescript-eslint/naming-convention
                    'Software Engineering': {
                        // eslint-disable-next-line @typescript-eslint/naming-convention
                        'Software Architect ': 0.7, // having space at the end of the key string
                        Developer: 0.9,
                    },
                },
            });

            // Act
            const result =
                await PairwiseScoreService.getWorkExpJobApplicationScore(
                    jobApplicationId,
                    mockRoleFitRecipeWithCompetencyInclusiveness,
                    mockRoleFitRecipeWithCompetencyInclusiveness.id,
                    5,
                );

            // Assert
            const matches = (
                result.score_outcome as PairwiseScoreOutcomePayload
            )?.pairwise_result[0]?.matches;
            expect(matches).toEqual(
                expect.arrayContaining([
                    {
                        job_title: 'Software Architect', // should not have space at the end
                        score: 0.7,
                    },
                    { job_title: 'Developer', score: 0.9 },
                ]),
            );
            expect(result).toMatchSnapshot();
        });

        it('should handle the case where the job application has no career', async () => {
            // Act
            const result =
                await PairwiseScoreService.getWorkExpJobApplicationScore(
                    '00000000-0000-0000-0001-000000000002',
                    mockRoleFitRecipeWithCompetencyInclusiveness,
                    mockRoleFitRecipeWithCompetencyInclusiveness.id,
                    5,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });

    describe('processPairwiseSkillOutput', () => {
        it('should return the correct pairwise skill output format with hard skill', async () => {
            // Act
            const result =
                await PairwiseScoreService.processPairwiseSkillOutput(
                    mockRoleFitRecipeWithCompetencyInclusiveness.job_competency,
                    ['Javascript', 'Python', 'Typescript '],
                    JobApplicationScoreType.HARD_SKILL,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should return the correct pairwise skill output format with work experience', async () => {
            // Act
            const result =
                await PairwiseScoreService.processPairwiseSkillOutput(
                    mockRoleFitRecipeWithCompetencyInclusiveness.job_competency,
                    ['Javascript', 'Python', 'Typescript'],
                    JobApplicationScoreType.WORK_EXP,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should handle the case where the job application has no career', async () => {
            // Act
            const result =
                await PairwiseScoreService.processPairwiseSkillOutput(
                    mockRoleFitRecipeWithCompetencyInclusiveness.job_competency,
                    [],
                    JobApplicationScoreType.WORK_EXP,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });

        it('should handle the case where the recipe have no skills', async () => {
            // Act
            const result =
                await PairwiseScoreService.processPairwiseSkillOutput(
                    [],
                    ['Javascript', 'Python', 'Typescript'],
                    JobApplicationScoreType.HARD_SKILL,
                );

            // Assert
            expect(result).toMatchSnapshot();
        });
    });
});
