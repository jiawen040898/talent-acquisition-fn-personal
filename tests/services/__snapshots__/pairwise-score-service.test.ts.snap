// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairwiseScoreService getHardSkillJobApplicationScore should handle the case where recipe have no skills 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 10,
  "score_dimension": 2,
  "score_outcome": {
    "pairwise_result": [],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "hard_skill",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService getHardSkillJobApplicationScore should handle the case where the job application has no skills 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "pairwise_result": [],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "hard_skill",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService getHardSkillJobApplicationScore should return the correct hard skill job application score 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 8,
  "score_dimension": 2,
  "score_outcome": {
    "pairwise_result": [
      {
        "matches": [
          {
            "match": false,
            "score": 0.7,
            "skill_name": "Python",
            "source": "daxtra",
          },
          {
            "match": true,
            "score": 0.9,
            "skill_name": "Typescript",
            "source": "daxtra",
          },
        ],
        "skill_name": "Javascript",
      },
    ],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "hard_skill",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService getWorkExpJobApplicationScore should handle the case where the job application has no career 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000002",
  "score": 0,
  "score_dimension": 0,
  "score_outcome": {
    "pairwise_result": [],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "work_experience",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService getWorkExpJobApplicationScore should return the correct work experience job application score 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 8,
  "score_dimension": 2,
  "score_outcome": {
    "pairwise_result": [
      {
        "matches": [
          {
            "job_title": "Python",
            "score": 0.7,
          },
          {
            "job_title": "Typescript",
            "score": 0.9,
          },
        ],
        "previous_employment": "Javascript",
      },
    ],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "work_experience",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService getWorkExpJobApplicationScore should trim job if the job key provided have space 1`] = `
{
  "created_by": 5,
  "job_application_id": "00000000-0000-0000-0001-000000000001",
  "score": 8,
  "score_dimension": 2,
  "score_outcome": {
    "pairwise_result": [
      {
        "matches": [
          {
            "job_title": "Software Architect",
            "score": 0.7,
          },
          {
            "job_title": "Developer",
            "score": 0.9,
          },
        ],
        "previous_employment": "Software Engineering",
      },
    ],
  },
  "score_recipe_id": "106e0b1d-02c4-4321-9dac-9768f84f8bc1",
  "score_type": "work_experience",
  "updated_by": 5,
}
`;

exports[`PairwiseScoreService processPairwiseSkillOutput should handle the case where the job application has no career 1`] = `
{
  "pairwiseOutcome": [],
  "pairwiseScore": 0,
}
`;

exports[`PairwiseScoreService processPairwiseSkillOutput should handle the case where the recipe have no skills 1`] = `
{
  "pairwiseOutcome": [],
  "pairwiseScore": 1,
}
`;

exports[`PairwiseScoreService processPairwiseSkillOutput should return the correct pairwise skill output format with hard skill 1`] = `
{
  "pairwiseOutcome": [
    {
      "matches": [
        {
          "score": 0.7,
          "skill_name": "Python",
        },
        {
          "score": 0.9,
          "skill_name": "Typescript",
        },
      ],
      "skill_name": "Javascript",
    },
  ],
  "pairwiseScore": 0.8,
}
`;

exports[`PairwiseScoreService processPairwiseSkillOutput should return the correct pairwise skill output format with work experience 1`] = `
{
  "pairwiseOutcome": [
    {
      "matches": [
        {
          "job_title": "Python",
          "score": 0.7,
        },
        {
          "job_title": "Typescript",
          "score": 0.9,
        },
      ],
      "previous_employment": "Javascript",
    },
  ],
  "pairwiseScore": 0.8,
}
`;
